src/v/archival/arrow_writer.cc

```diff
@@ -0,0 +1,323 @@
+
+        if (record.has_value()) {
+            value = iobuf_to_string(record.value());
+        }
+
+        _ok = key_builder.Append(key);
+        if (!_ok.ok()) {
+            return ss::stop_iteration::yes;
+        }
+
+        _ok = value_builder.Append(value);
+        if (!_ok.ok()) {
+            return ss::stop_iteration::yes;
+        }
+
+        _ok = timestamp_builder.Append(
+          batch.header().first_timestamp.value() + record.timestamp_delta());
+        if (!_ok.ok()) {
+            return ss::stop_iteration::yes;
+        }
+
+        // FIXME(jcipar): this is not the correct way to compute offsets. The
+        // new log reader config has a `translate_offsets` options that
+        // automatically does the translation. After rebasing, use that.
+        _ok = offset_builder.Append(
+          int64_t(batch.header().base_offset) + record.offset_delta());
+        if (!_ok.ok()) {
+            return ss::stop_iteration::yes;
+        }
+
+        _rows++;
+        return ss::stop_iteration::no;
+    });
+    if (!_ok.ok()) {
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::yes);
+    }
+
+    std::shared_ptr<arrow::Array> key_array;
+    std::shared_ptr<arrow::Array> value_array;
+    std::shared_ptr<arrow::Array> offset_array;
+    std::shared_ptr<arrow::Array> timestamp_array;
+
+    auto&& key_builder_result = key_builder.Finish();
+    // Arrow ASSIGN_OR_RAISE macro doesn't actually raise, it returns a
+    // not-ok value. Expanding the macro and editing ensures we get the
+    // correct return type.
+    _ok = key_builder_result.status();
+    if (!_ok.ok()) {
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::yes);
+    } else {
+        key_array = std::move(key_builder_result).ValueUnsafe();
+    }
+
+    auto&& value_builder_result = value_builder.Finish();
+    _ok = value_builder_result.status();
+    if (!_ok.ok()) {
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::yes);
+    } else {
+        value_array = std::move(value_builder_result).ValueUnsafe();
+    }
+
+    auto&& offset_builder_result = offset_builder.Finish();
+    _ok = offset_builder_result.status();
+    if (!_ok.ok()) {
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::yes);
+    } else {
+        offset_array = std::move(offset_builder_result).ValueUnsafe();
+    }
+
+    auto&& timestamp_builder_result = timestamp_builder.Finish();
+    _ok = timestamp_builder_result.status();
+    if (!_ok.ok()) {
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::yes);
+    } else {
+        timestamp_array = std::move(timestamp_builder_result).ValueUnsafe();
+    }
+
+    _key_vector.push_back(key_array);
+    _value_vector.push_back(value_array);
+    _offset_vector.push_back(offset_array);
+    _timestamp_vector.push_back(timestamp_array);
```

Comment:
Either one of these can throw, potentially leaving you inconsistent.

You can avoid that by calling `vec.reserve(vec.size() +1);` on each of them before mutating anything.